File: android/app/src/main/java/com/tinyswords/app/game/Upgrades.kt
package com.tinyswords.app.game

import com.tinyswords.app.game.entities.GameBuilding
import com.tinyswords.app.game.entities.GameUnit
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

data class UpgradeCost(val wood: Int, val gold: Int, val food: Int = 0)

data class BuildingUpgradeSpec(
    val maxLevel: Int,
    val costs: List<UpgradeCost>,
    val hpMult: List<Float> = listOf(1f),
    val popBonus: List<Int> = listOf(0),
    val trainMult: List<Float> = listOf(1f),
    val rangeBonus: List<Float> = listOf(0f),
    val damageBonus: List<Int> = listOf(0),
    val archerCount: List<Int> = listOf(0),
    val workerHpBonus: List<Int> = listOf(0),
    val workerDamageBonus: List<Int> = listOf(0),
    val workerSpeedBonus: List<Float> = listOf(0f),
    val archerHpBonus: List<Int> = listOf(0),
    val archerDamageBonus: List<Int> = listOf(0),
    val archerRangeBonus: List<Float> = listOf(0f)
)

val BUILDING_UPGRADE_SPECS = mapOf(
    "tower" to BuildingUpgradeSpec(
        maxLevel = 2,
        costs = listOf(UpgradeCost(120, 135)),
        hpMult = listOf(1.00f, 2.05f),
        rangeBonus = listOf(0f, 60f),
        damageBonus = listOf(0, 4),
        archerCount = listOf(1, 2)
    ),
    "castle" to BuildingUpgradeSpec(
        maxLevel = 4,
        costs = listOf(UpgradeCost(180, 140), UpgradeCost(260, 240), UpgradeCost(360, 340)),
        hpMult = listOf(1.00f, 1.22f, 1.48f, 1.78f),
        popBonus = listOf(0, 4, 8, 12),
        trainMult = listOf(1.00f, 0.90f, 0.80f, 0.70f),
        rangeBonus = listOf(300f, 335f, 370f, 405f),
        damageBonus = listOf(0, 2, 5, 8),
        archerCount = listOf(1, 2, 3, 3)
    ),
    "house" to BuildingUpgradeSpec(
        maxLevel = 2,
        costs = listOf(UpgradeCost(95, 55)),
        hpMult = listOf(1.00f, 1.70f),
        popBonus = listOf(0, 6),
        workerHpBonus = listOf(0, 14),
        workerDamageBonus = listOf(0, 2),
        workerSpeedBonus = listOf(0f, 8f)
    ),
    "barracks" to BuildingUpgradeSpec(
        maxLevel = 2,
        costs = listOf(UpgradeCost(150, 120)),
        hpMult = listOf(1.00f, 1.40f),
        trainMult = listOf(1.00f, 0.78f)
    ),
    "archery" to BuildingUpgradeSpec(
        maxLevel = 2,
        costs = listOf(UpgradeCost(145, 135)),
        hpMult = listOf(1.00f, 1.35f),
        trainMult = listOf(1.00f, 0.78f),
        archerHpBonus = listOf(0, 16),
        archerDamageBonus = listOf(0, 4),
        archerRangeBonus = listOf(0f, 38f)
    ),
    "monastery" to BuildingUpgradeSpec(
        maxLevel = 2,
        costs = listOf(UpgradeCost(135, 145)),
        hpMult = listOf(1.00f, 1.35f),
        trainMult = listOf(1.00f, 0.82f)
    )
)

fun upgradeSpec(type: String): BuildingUpgradeSpec? = BUILDING_UPGRADE_SPECS[type]
fun buildingUpgradeMaxLevel(type: String): Int = upgradeSpec(type)?.maxLevel ?: 1
fun upgradeLevel(building: GameBuilding): Int = building.level.coerceIn(1, buildingUpgradeMaxLevel(building.type))
fun upgradeCostFor(building: GameBuilding): UpgradeCost? {
    val spec = upgradeSpec(building.type) ?: return null
    val currentLevel = upgradeLevel(building)
    if (currentLevel >= spec.maxLevel) return null
    return spec.costs.getOrNull(currentLevel - 1)
}

private fun <T> levelValue(values: List<T>, level: Int, fallback: T): T {
    if (values.isEmpty()) return fallback
    val index = (level - 1).coerceIn(0, values.lastIndex)
    return values[index]
}

fun buildingMaxHpFor(type: String, level: Int = 1): Int {
    val base = BUILDINGS[type]?.hp ?: 100
    val mult = levelValue(upgradeSpec(type)?.hpMult ?: listOf(1f), level, 1f)
    return max(1, (base * mult).roundToInt())
}

fun normalizeBuildingStats(building: GameBuilding, preserveRatio: Boolean = true) {
    val oldMax = max(1, building.maxHp)
    val oldHp = building.hp.coerceIn(0, oldMax)
    building.level = upgradeLevel(building)
    val nextMax = buildingMaxHpFor(building.type, building.level)
    val wasStructurallyFull = building.buildProgress >= 1f && oldHp >= oldMax - 1
    building.maxHp = nextMax
    building.hp = if (preserveRatio) {
        val ratio = oldHp.toFloat() / oldMax.toFloat()
        (nextMax * ratio).roundToInt().coerceIn(if (building.buildProgress < 1f) 0 else 1, nextMax)
    } else if (wasStructurallyFull && nextMax > oldMax) {
        nextMax
    } else {
        oldHp.coerceIn(0, nextMax)
    }
}

fun buildingPopulationCapacity(building: GameBuilding): Int {
    if (building.dead || building.buildProgress < 1f) return 0
    val def = BUILDINGS[building.type] ?: return 0
    val bonus = levelValue(upgradeSpec(building.type)?.popBonus ?: listOf(0), upgradeLevel(building), 0)
    return def.pop + bonus
}

fun buildingTrainTime(building: GameBuilding, unitType: String): Float {
    val base = UNITS[unitType]?.trainTime ?: 1f
    val mult = levelValue(upgradeSpec(building.type)?.trainMult ?: listOf(1f), upgradeLevel(building), 1f)
    return max(1f, base * mult)
}

fun defensiveArcherCount(building: GameBuilding): Int {
    if (building.dead || building.buildProgress < 1f) return 0
    val fallback = if (BUILDINGS[building.type]?.builtInArcher == true) 1 else 0
    return levelValue(upgradeSpec(building.type)?.archerCount ?: listOf(fallback), upgradeLevel(building), fallback)
}

fun defensiveBuildingRange(building: GameBuilding): Float {
    if (defensiveArcherCount(building) <= 0) return 0f
    val rangeBonus = levelValue(upgradeSpec(building.type)?.rangeBonus ?: listOf(0f), upgradeLevel(building), 0f)
    return if (building.type == "tower") (BUILDINGS["tower"]?.towerRange ?: 0f) + rangeBonus else rangeBonus
}

fun defensiveBuildingDamage(building: GameBuilding): Int {
    val base = UNITS["archer"]?.damage ?: 12
    val bonus = levelValue(upgradeSpec(building.type)?.damageBonus ?: listOf(0), upgradeLevel(building), 0)
    return base + bonus
}

fun defensiveBuildingCooldown(building: GameBuilding): Float {
    val level = upgradeLevel(building)
    val base = if (building.type == "tower") 1.18f * 0.92f else 1.18f * 1.08f
    val levelHaste = if (building.type == "castle") max(0, level - 1) * 0.04f else 0f
    return max(0.72f, base - levelHaste)
}

data class UnitUpgradeBonuses(
    val workerHp: Int = 0,
    val workerDamage: Int = 0,
    val workerSpeed: Float = 0f,
    val archerHp: Int = 0,
    val archerDamage: Int = 0,
    val archerRange: Float = 0f
)

fun factionUnitUpgradeBonuses(state: GameState, factionId: Int): UnitUpgradeBonuses {
    var workerHp = 0
    var workerDamage = 0
    var workerSpeed = 0f
    var archerHp = 0
    var archerDamage = 0
    var archerRange = 0f
    for (b in state.buildings) {
        if (b.dead || b.faction != factionId || b.buildProgress < 1f) continue
        val spec = upgradeSpec(b.type) ?: continue
        val level = upgradeLevel(b)
        workerHp = max(workerHp, levelValue(spec.workerHpBonus, level, 0))
        workerDamage = max(workerDamage, levelValue(spec.workerDamageBonus, level, 0))
        workerSpeed = max(workerSpeed, levelValue(spec.workerSpeedBonus, level, 0f))
        archerHp = max(archerHp, levelValue(spec.archerHpBonus, level, 0))
        archerDamage = max(archerDamage, levelValue(spec.archerDamageBonus, level, 0))
        archerRange = max(archerRange, levelValue(spec.archerRangeBonus, level, 0f))
    }
    return UnitUpgradeBonuses(workerHp, workerDamage, workerSpeed, archerHp, archerDamage, archerRange)
}

fun applyFactionUnitUpgrades(state: GameState, unit: GameUnit, preserveRatio: Boolean = true) {
    val def = UNITS[unit.type] ?: return
    val oldMax = max(1, unit.maxHp)
    val oldHp = unit.hp.coerceIn(0, oldMax)
    val bonus = factionUnitUpgradeBonuses(state, unit.faction)
    var nextMax = def.hp
    var nextDamage = def.damage
    var nextSpeed = def.speed
    var nextRange = def.range
    when (unit.type) {
        "worker" -> {
            nextMax += bonus.workerHp
            nextDamage += bonus.workerDamage
            nextSpeed += bonus.workerSpeed
        }
        "archer" -> {
            nextMax += bonus.archerHp
            nextDamage += bonus.archerDamage
            nextRange += bonus.archerRange
        }
    }
    unit.maxHp = nextMax
    unit.damage = nextDamage
    unit.speed = nextSpeed
    unit.range = nextRange
    unit.maxCd = def.cd
    unit.hp = if (preserveRatio) {
        val ratio = oldHp.toFloat() / oldMax.toFloat()
        (nextMax * ratio).roundToInt().coerceIn(1, nextMax)
    } else {
        oldHp.coerceIn(1, nextMax)
    }
}

fun applyFactionUnitUpgradesToAll(state: GameState, factionId: Int) {
    for (unit in state.units) if (!unit.dead && unit.faction == factionId) applyFactionUnitUpgrades(state, unit)
}

File: js/core/upgrades.js
// Building upgrades and derived combat/economy stats.
'use strict';

const BUILDING_UPGRADE_SPECS = Object.freeze({
  tower: {
    maxLevel: 2,
    costs: [null, { wood: 120, gold: 135, food: 0 }],
    hpMult: [1.00, 2.05],
    rangeBonus: [0, 60],
    damageBonus: [0, 4],
    archerCount: [1, 2]
  },
  castle: {
    maxLevel: 4,
    costs: [null, { wood: 180, gold: 140, food: 0 }, { wood: 260, gold: 240, food: 0 }, { wood: 360, gold: 340, food: 0 }],
    hpMult: [1.00, 1.22, 1.48, 1.78],
    popBonus: [0, 4, 8, 12],
    trainMult: [1.00, 0.90, 0.80, 0.70],
    rangeBonus: [300, 335, 370, 405],
    damageBonus: [0, 2, 5, 8],
    archerCount: [1, 2, 3, 3]
  },
  house: {
    maxLevel: 2,
    costs: [null, { wood: 95, gold: 55, food: 0 }],
    hpMult: [1.00, 1.70],
    popBonus: [0, 6],
    workerHpBonus: [0, 14],
    workerDamageBonus: [0, 2],
    workerSpeedBonus: [0, 8]
  },
  barracks: {
    maxLevel: 2,
    costs: [null, { wood: 150, gold: 120, food: 0 }],
    hpMult: [1.00, 1.40],
    trainMult: [1.00, 0.78]
  },
  archery: {
    maxLevel: 2,
    costs: [null, { wood: 145, gold: 135, food: 0 }],
    hpMult: [1.00, 1.35],
    trainMult: [1.00, 0.78],
    archerHpBonus: [0, 16],
    archerDamageBonus: [0, 4],
    archerRangeBonus: [0, 38]
  },
  monastery: {
    maxLevel: 2,
    costs: [null, { wood: 135, gold: 145, food: 0 }],
    hpMult: [1.00, 1.35],
    trainMult: [1.00, 0.82]
  }
});

function levelValue(list, level, fallback) {
  if (!Array.isArray(list) || !list.length) return fallback;
  const idx = clamp((level || 1) - 1, 0, list.length - 1);
  return list[idx] === undefined ? fallback : list[idx];
}

function buildingUpgradeSpec(type) { return BUILDING_UPGRADE_SPECS[type] || null; }
function buildingUpgradeMaxLevel(type) { return (buildingUpgradeSpec(type) && buildingUpgradeSpec(type).maxLevel) || 1; }
function buildingLevel(building) { return clamp(Number(building && building.level) || 1, 1, buildingUpgradeMaxLevel(building && building.type)); }
function buildingUpgradeCost(building) {
  const spec = buildingUpgradeSpec(building && building.type);
  if (!spec) return null;
  const current = buildingLevel(building);
  if (current >= spec.maxLevel) return null;
  return spec.costs[current] || null;
}
function buildingUpgradeLabel(building) {
  const next = buildingLevel(building) + 1;
  return `${BUILDINGS[building.type].label} Lv.${next}`;
}
function buildingMaxHpFor(type, level = 1) {
  const def = BUILDINGS[type];
  if (!def) return 1;
  const spec = buildingUpgradeSpec(type);
  const mult = levelValue(spec && spec.hpMult, level, 1);
  return Math.max(1, Math.round(def.hp * mult));
}
function normalizeBuildingStats(building, preserveRatio = true) {
  if (!building || !BUILDINGS[building.type]) return building;
  const oldMax = Math.max(1, Number(building.maxHp) || BUILDINGS[building.type].hp);
  const oldHp = clamp(Number(building.hp) || oldMax, 0, oldMax);
  building.level = buildingLevel(building);
  const nextMax = buildingMaxHpFor(building.type, building.level);
  const wasStructurallyFull = building.build >= 1 && oldHp >= oldMax - 1;
  building.maxHp = nextMax;
  if (preserveRatio) {
    const ratio = oldMax > 0 ? clamp(oldHp / oldMax, 0, 1) : 1;
    building.hp = clamp(Math.max(building.build < 1 ? oldHp : 1, Math.round(nextMax * ratio)), 0, nextMax);
  } else if (wasStructurallyFull && nextMax > oldMax) {
    // Old saves and level migrations often have hp equal to the old max. Treat
    // that as healthy rather than damaged, otherwise every upgraded building
    // renders a fake floating HP bar forever.
    building.hp = nextMax;
  } else {
    building.hp = clamp(oldHp, 0, nextMax);
  }
  return building;
}
function buildingPopulationCapacity(building) {
  if (!building || building.dead || building.build < 1) return 0;
  const def = BUILDINGS[building.type];
  const spec = buildingUpgradeSpec(building.type);
  return (def && def.pop || 0) + levelValue(spec && spec.popBonus, buildingLevel(building), 0);
}
function buildingTrainTime(building, unitType) {
  const unit = UNITS[unitType];
  if (!unit) return 1;
  const spec = buildingUpgradeSpec(building && building.type);
  const mult = levelValue(spec && spec.trainMult, buildingLevel(building), 1);
  return Math.max(1, unit.time * mult);
}
function defensiveArcherCount(building) {
  if (!building || building.dead || building.build < 1) return 0;
  const spec = buildingUpgradeSpec(building.type);
  return levelValue(spec && spec.archerCount, buildingLevel(building), BUILDINGS[building.type] && BUILDINGS[building.type].builtInArcher ? 1 : 0) || 0;
}
function defensiveBuildingRange(building) {
  if (!defensiveArcherCount(building)) return 0;
  const spec = buildingUpgradeSpec(building.type);
  if (building.type === 'tower') return (BUILDINGS.tower.range || 0) + levelValue(spec && spec.rangeBonus, buildingLevel(building), 0);
  return levelValue(spec && spec.rangeBonus, buildingLevel(building), 0);
}
function defensiveBuildingDamage(building) {
  const spec = buildingUpgradeSpec(building && building.type);
  return (UNITS.archer.damage || 0) + levelValue(spec && spec.damageBonus, buildingLevel(building), 0);
}
function defensiveBuildingCooldown(building) {
  // Each defensive archer fires one projectile per volley. Do not shorten the
  // cooldown by archer count, otherwise high-level castles multiply damage twice
  // and become unreadable. Damage scales through projectile count + upgrades.
  const level = buildingLevel(building);
  const base = building && building.type === 'tower' ? UNITS.archer.cd * 0.92 : UNITS.archer.cd * 1.08;
  const levelHaste = building && building.type === 'castle' ? Math.max(0, level - 1) * 0.04 : 0;
  return Math.max(0.72, base - levelHaste);
}

function factionUnitUpgradeBonuses(game, factionId) {
  const bonus = { workerHp: 0, workerDamage: 0, workerSpeed: 0, archerHp: 0, archerDamage: 0, archerRange: 0 };
  if (!game) return bonus;
  for (const b of game.buildings || []) {
    if (!b || b.dead || b.faction !== factionId || b.build < 1) continue;
    const spec = buildingUpgradeSpec(b.type);
    if (!spec) continue;
    const level = buildingLevel(b);
    bonus.workerHp = Math.max(bonus.workerHp, levelValue(spec.workerHpBonus, level, 0));
    bonus.workerDamage = Math.max(bonus.workerDamage, levelValue(spec.workerDamageBonus, level, 0));
    bonus.workerSpeed = Math.max(bonus.workerSpeed, levelValue(spec.workerSpeedBonus, level, 0));
    bonus.archerHp = Math.max(bonus.archerHp, levelValue(spec.archerHpBonus, level, 0));
    bonus.archerDamage = Math.max(bonus.archerDamage, levelValue(spec.archerDamageBonus, level, 0));
    bonus.archerRange = Math.max(bonus.archerRange, levelValue(spec.archerRangeBonus, level, 0));
  }
  return bonus;
}
function unitDerivedStats(game, unit) {
  const def = UNITS[unit && unit.type] || UNITS.worker;
  const bonus = factionUnitUpgradeBonuses(game, unit && unit.faction);
  const stats = { maxHp: def.hp, speed: def.speed, range: def.range, damage: def.damage, cd: def.cd };
  if (unit && unit.type === 'worker') {
    stats.maxHp += bonus.workerHp;
    stats.damage += bonus.workerDamage;
    stats.speed += bonus.workerSpeed;
  } else if (unit && unit.type === 'archer') {
    stats.maxHp += bonus.archerHp;
    stats.damage += bonus.archerDamage;
    stats.range += bonus.archerRange;
  }
  return stats;
}
function normalizeUnitStats(game, unit, preserveRatio = true) {
  if (!unit || !UNITS[unit.type]) return unit;
  const oldMax = Number(unit.maxHp) || UNITS[unit.type].hp;
  const oldHp = Number(unit.hp) || oldMax;
  const stats = unitDerivedStats(game, unit);
  unit.maxHp = stats.maxHp;
  unit.speed = stats.speed;
  unit.range = stats.range;
  unit.damage = stats.damage;
  if (preserveRatio) {
    const ratio = oldMax > 0 ? clamp(oldHp / oldMax, 0, 1) : 1;
    unit.hp = clamp(Math.max(1, Math.round(stats.maxHp * ratio)), 0, stats.maxHp);
  } else {
    unit.hp = clamp(oldHp, 0, stats.maxHp);
  }
  return unit;
}
function unitCombatRange(game, unit) { return unitDerivedStats(game, unit).range; }
function unitCombatDamage(game, unit) { return unitDerivedStats(game, unit).damage; }
function unitAttackCooldown(game, unit) { return unitDerivedStats(game, unit).cd; }

function upgradeBuildingForPlayer(game, building) {
  if (!game || !building || building.entity !== 'building' || building.faction !== 0 || building.build < 1) return false;
  const cost = buildingUpgradeCost(building);
  if (!cost) { game.toast('Already fully upgraded.', 1.2); game.sfx.deny(); return false; }
  if (!pay(game.factions[0], cost)) { game.toast(`Upgrade needs ${fmtCost(cost)}.`, 1.5); game.sfx.deny(); return false; }
  building.level = buildingLevel(building) + 1;
  normalizeBuildingStats(building, false);
  // Upgrades are a full structural refit; fill HP to the new max so the world
  // renderer does not leave unwanted floating HP bars after a clean upgrade.
  building.hp = building.maxHp;
  for (const unit of game.units) if (unit.faction === building.faction && !unit.dead) normalizeUnitStats(game, unit, true);
  game.effects.push({ kind: 'upgrade', x: building.x, y: building.y - building.h * .45, time: .9, max: .9 });
  game.toast(`${BUILDINGS[building.type].label} upgraded to Lv.${building.level}.`, 1.6);
  game.sfx.build(game.audioGainAt ? game.audioGainAt(building.x, building.y) : 1);
  game.uiDirty = true;
  return true;
}
